pkg/cluster/cluster_info_scraper.go

```diff
@@ -283,36 +241,85 @@ func (s *ClusterScraper) GetPVCs(namespace string, opts metav1.ListOptions) ([]*
 	return pvcs, nil
 }
 
-// GetPodGrandparentInfo gets grandParent (parent's parent) information of a pod: kind, name, uid
+// UpdatePodControllerCache is called at the beginning of a full discovery after all pods and a predefined types of
+// workload controllers (see processor.ControllerProcessor for details) are discovered. We tried to cache all pods'
+// controller info (i.e., kind, name, uid) into the cache maintained by ClusterScraper for speedy look up when building
+// entity DTOs.
+// A pod can have its labels changed on the fly without being restarted, causing its controller to be changed as well.
+// This is very unlikely but still possible. This can be detected when the cache entry expires in defaultCacheTTL.
+func (s *ClusterScraper) UpdatePodControllerCache(
+	pods []*api.Pod, controllers map[string]*repository.K8sController) {
+	existing := 0
+	added := 0
+	for _, pod := range pods {
+		podControllerInfoKey := util.PodControllerInfoKey(pod)
+		if _, exists := s.cache.Get(podControllerInfoKey); exists {
+			// Pod's controller info already exists in the cache
+			existing++
+			continue
+		}
+		ownerInfo, err := util.GetPodParentInfo(pod)
+		if err != nil || util.IsOwnerInfoEmpty(ownerInfo) {
+			// Pod does not have controller
+			glog.V(3).Infof("Skip updating controller for pod %v/%v: pod has no controller.",
+				pod.Namespace, pod.Name)
+			continue
+		}
```

Comment:
Nit: can we introduce a new variable to record the count of bare pods?

Then the numbers will match in the log you added at the end of this function:
```
total number of pods = existing + added + bare pods + skipped pods (with custom controller)
```